File: DailyChallenges/CPP/LeetCode_2246.cpp
class Solution {
public:
    int longestPath(vector<int>& parent, string s) {
        int n = parent.size();

        unordered_map<int, vector<int>> m; 
        for (int i = 1;i < n; ++i)
            m[parent[i]].push_back(i);

        int ans = 1;
        queue<int> q; q.push(0);
        
        function<int(int)> dfs = [&](int node) {
            if (m[node].size() == 0) return 1; // reach the end
            vector<int> chLen;
            for (int child: m[node]){
                if (s[child] == s[node])
                    q.push(child);
                else 
                    chLen.push_back(dfs(child));
            }
            int cl_n = chLen.size();
            if (cl_n > 1){
                nth_element(chLen.begin(), chLen.end()-2, chLen.end());
                ans = max(ans, chLen[cl_n-1] + chLen[cl_n-2] + 1);
                return chLen[cl_n-1] + 1;
            }else if (cl_n == 1){
                ans = max(ans, chLen[0] + 1);
                return chLen[0] + 1;
            }else
                return 1;
        };
        
        while (!q.empty()){
            int node = q.front(); q.pop();
            dfs(node);
        }

        return ans;
    }
};

File: DailyChallenges/Scala/LeetCode_2246.scala
object Solution {
    def longestPath(parent: Array[Int], s: String): Int = {
        // Standard process to prepare a map of children
        val m = scala.collection.mutable.Map.empty[Int, List[Int]]
        for (i <- 1 until parent.length)
            m(parent(i)) = i :: m.getOrElse(parent(i), Nil)
        
        // When two adjacent nodes share same label, it break the edge actually.
        // can be treated as a seperate tree root at child node.
        val q = scala.collection.mutable.Queue.empty[Int]
        q.enqueue(0)
        
        var ans = 1;
        def dfs(node: Int): Int = {
            if (!m.contains(node)) return 1 // leaf
            val big2 = Array(0, 0, 0)
            for (ch <- m(node)){
                if (s(node) == s(ch)) q.enqueue(ch) else {
                    val cl = dfs(ch)
                    if (cl > big2(1)) big2(1) = cl
                    if (big2(1) > big2(0)){ // swap 1st and 2nd using 3rd as a tmp
                        big2(2) = big2(1); big2(1) = big2(0); big2(0) = big2(2)
                    }
                }
            }
            ans = Math.max(ans, big2(0)+big2(1)+1)
            big2(0)+1
        }

        while (q.nonEmpty) dfs(q.dequeue)
        
        ans
    }
}
